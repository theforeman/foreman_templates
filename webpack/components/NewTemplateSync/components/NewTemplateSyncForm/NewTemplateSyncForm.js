import React from 'react';
import PropTypes from 'prop-types';

import ForemanForm from 'foremanReact/components/common/forms/ForemanForm';
import * as Yup from 'yup';
import SyncSettingsFields from '../SyncSettingFields';
import SyncTypeRadios from '../SyncTypeRadios';

const redirectToResult = history => () =>
  history.push({ pathname: '/template_syncs/result' });

const repoFormat = formatAry => value => {
  if (value === undefined) {
    return true;
  }

  const valid = formatAry
    .map(item => value.startsWith(item))
    .reduce((memo, item) => item || memo, false);

  return value && valid;
};

const syncFormSchema = (syncType, settingsObj, validationData) => {
  const schema = (settingsObj[syncType].asMutable() || []).reduce(
    (memo, setting) => {
      if (setting.name === 'repo') {
        return {
          ...memo,
          repo: Yup.string()
            .test(
              'repo-format',
              `Invalid repo format, must start with one of: ${validationData.repo.join(
                ', '
              )}`,
              repoFormat(validationData.repo)
            )
            .required("can't be blank"),
        };
      }
      return memo;
    },
    {}
  );

  return Yup.object().shape({
    [syncType]: Yup.object().shape(schema),
  });
};

class NewTemplateSyncForm extends React.Component {
  allowedSyncType = (userPermissions, radioAttrs) =>
    this.props.userPermissions[radioAttrs.permission];

  constructor(props) {
    super(props);

    this.radioButtons = [
      { label: 'Import', value: 'import', permission: 'import' },
      { label: 'Export', value: 'export', permission: 'export' },
    ];

    this.state = {
      syncType: this.radioButtons.find(radioAttrs =>
        this.allowedSyncType(props.userPermissions, radioAttrs)
      ).value,
    };
  }

  updateSyncType = event => {
    this.setState({ syncType: event.target.value });
  };

  permitRadioButtons = buttons =>
    buttons.filter(buttonAttrs =>
      this.allowedSyncType(this.props.userPermissions, buttonAttrs)
    );

  initRadioButtons = syncType =>
    this.permitRadioButtons(this.radioButtons).map(buttonAttrs => ({
      get checked() {
        return this.value === syncType;
      },
      onChange: this.updateSyncType,
      ...buttonAttrs,
    }));

  render() {
    const {
      error,
      submitForm,
      importSettings,
      exportSettings,
      history,
      validationData,
      importUrl,
      exportUrl,
      initialValues,
    } = this.props;

    const resetToDefault = (fieldName, fieldValue) => resetFn =>
      resetFn(fieldName, fieldValue);

    return (
      <ForemanForm
        onSubmit={(values, actions) => {
          const url = this.state.syncType === 'import' ? importUrl : exportUrl;
          return submitForm({
            url,
            values: values[this.state.syncType],
            message: `Templates were ${this.state.syncType}ed.`,
            item: 'TemplateSync',
          }).then(args => {
            history.replace({ pathname: '/template_syncs/result' });
          });
        }}
        initialValues={initialValues}
        validationSchema={syncFormSchema(
          this.state.syncType,
          { import: importSettings, export: exportSettings },
          validationData
        )}
        onCancel={redirectToResult(history)}
        error={error}
      >
        <SyncTypeRadios
          name="syncType"
          controlLabel="Action type"
          radios={this.initRadioButtons(this.state.syncType)}
        />
        <SyncSettingsFields
          importSettings={importSettings}
          exportSettings={exportSettings}
          syncType={this.state.syncType}
          resetField={resetToDefault}
        />
      </ForemanForm>
    );
  }
}

NewTemplateSyncForm.propTypes = {
  importSettings: PropTypes.array,
  exportSettings: PropTypes.array,
  userPermissions: PropTypes.object.isRequired,
  error: PropTypes.object,
  history: PropTypes.object,
  validationData: PropTypes.object,
  initialValues: PropTypes.object.isRequired,
  exportUrl: PropTypes.string.isRequired,
  importUrl: PropTypes.string.isRequired,
  submitForm: PropTypes.func.isRequired,
};

NewTemplateSyncForm.defaultProps = {
  importSettings: [],
  exportSettings: [],
  validationData: {},
  error: undefined,
  history: {},
};

export default NewTemplateSyncForm;
