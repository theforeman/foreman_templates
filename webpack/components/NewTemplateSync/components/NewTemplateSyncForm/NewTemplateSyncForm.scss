#sync-form {

  &.pf-v5-c-form {
    --pf-v5-c-form--GridGap: 0
  }
}