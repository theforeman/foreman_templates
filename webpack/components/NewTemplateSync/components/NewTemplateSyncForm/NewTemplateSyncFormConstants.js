export const NEW_TEMPLATE_SYNC_FORM_NAME = 'newTemplateSync';
