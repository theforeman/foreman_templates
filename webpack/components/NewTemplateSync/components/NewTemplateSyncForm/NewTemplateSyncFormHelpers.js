import * as Yup from 'yup';

export const redirectToResult = history => () =>
  history.push({ pathname: '/template_syncs/result' });

export const repoFormat = formatAry => value => {
  if (value === undefined) {
    return true;
  }

  const valid = formatAry
    .map(item => value.startsWith(item))
    .reduce((memo, item) => item || memo, false);

  return value && valid;
};

export const syncFormSchema = (syncType, settingsObj, validationData) => {
  const schema = (settingsObj[syncType].asMutable() || []).reduce(
    (memo, setting) => {
      if (setting.name === 'repo') {
        return {
          ...memo,
          repo: Yup.string()
            .test(
              'repo-format',
              `Invalid repo format, must start with one of: ${validationData.repo.join(
                ', '
              )}`,
              repoFormat(validationData.repo)
            )
            .required("can't be blank"),
        };
      }
      return memo;
    },
    {}
  );

  return Yup.object().shape({
    [syncType]: Yup.object().shape(schema),
  });
};
