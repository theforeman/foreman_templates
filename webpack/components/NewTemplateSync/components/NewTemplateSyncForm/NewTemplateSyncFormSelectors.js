import { createSelector } from 'reselect';

import {
  selectImportSettings,
  selectExportSettings,
} from '../../NewTemplateSyncSelectors';

export const transformInitialValues = settings =>
  settings.reduce(
    (memo, item) => Object.assign(memo, { [item.name]: item.value }),
    {}
  );

export const selectInitialFormValues = createSelector(
  selectImportSettings,
  selectExportSettings,
  (importSettings, exportSettings) => ({
    import: transformInitialValues(importSettings),
    export: transformInitialValues(exportSettings),
  })
);
