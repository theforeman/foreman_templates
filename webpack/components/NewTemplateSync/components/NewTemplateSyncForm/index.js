import { connect } from 'react-redux';
import { reduxForm } from 'redux-form';

import * as FormActions from 'foremanReact/redux/actions/common/forms';

import { NEW_TEMPLATE_SYNC_FORM_NAME } from './NewTemplateSyncFormConstants';
import NewTemplateSyncForm from './NewTemplateSyncForm';

import {
  selectImportSettings,
  selectExportSettings,
} from '../../NewTemplateSyncSelectors';
import {
  selectInitialFormValues,
  selectRegisteredFields,
} from './NewTemplateSyncFormSelectors';

const mapStateToProps = (state, ownProps) => {
  const importSettings = selectImportSettings(state);

  const exportSettings = selectExportSettings(state);

  const initialFormValues = selectInitialFormValues(state);

  const currentFields = selectRegisteredFields(
    NEW_TEMPLATE_SYNC_FORM_NAME,
    state
  );

  return {
    initialValues: { ...initialFormValues },
    importSettings,
    exportSettings,
    currentFields,
  };
};

const form = reduxForm({ form: NEW_TEMPLATE_SYNC_FORM_NAME })(
  NewTemplateSyncForm
);
export default connect(mapStateToProps, FormActions)(form);
