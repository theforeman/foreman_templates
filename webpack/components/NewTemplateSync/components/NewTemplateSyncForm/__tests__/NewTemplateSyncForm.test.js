import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import NewTemplateSyncForm from '../NewTemplateSyncForm';

import {
  importSettings,
  exportSettings,
} from '../../../__fixtures__/templateSyncSettings.fixtures';

const noop = () => {};

const commonFixture = {
  importUrl: '/import',
  exportUrl: '/export',
  validationData: {},
  userPermissions: {
    import: true,
    export: true,
  },
  handleSubmit: noop,
  valid: true,
  importSettings,
  exportSettings,
};

const fixtures = {
  'should render when for import settings': {
    loadingSettings: false,
    ...commonFixture,
  },
  'should render for export settings': {
    loadingSettings: true,
    ...commonFixture,
    userPermissions: {
      import: false,
      export: true,
    },
  },
};

describe('NewTemplateSyncForm', () =>
  testComponentSnapshotsWithFixtures(NewTemplateSyncForm, fixtures));
