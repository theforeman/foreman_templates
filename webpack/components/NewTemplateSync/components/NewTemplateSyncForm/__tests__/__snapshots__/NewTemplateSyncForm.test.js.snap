// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTemplateSyncForm should render for export settings 1`] = `
<Form
  disabled={false}
  errorTitle="Warning! "
  onCancel={[Function]}
  submitting={false}
>
  <SyncTypeRadios
    controlLabel="Action type"
    disabled={false}
    name="syncType"
    radios={
      Array [
        Object {
          "checked": true,
          "label": "Export",
          "onChange": [Function],
          "permission": "export",
          "value": "export",
        },
      ]
    }
  />
  <SyncSettingsFields
    disabled={false}
    exportSettings={
      Array [
        Object {
          "description": "Import or export names matching this regex (case-insensitive; snippets are not filtered)",
          "id": 47,
          "name": "filter",
          "settingsType": "string",
          "value": "",
        },
        Object {
          "description": "Negate the filter for import/export",
          "id": 48,
          "name": "negate",
          "settingsType": "bool",
          "value": false,
        },
        Object {
          "description": "Target path to import and export. Different protocols can be used, for example /tmp/dir, git://example.com, https://example.com, ssh://example.com. When exporting to /tmp, note that production deployments may be configured to use private tmp.",
          "id": 49,
          "name": "repo",
          "settingsType": "string",
          "value": "https://github.com/theforeman/community-templates.git",
        },
      ]
    }
    importSettings={
      Array [
        Object {
          "description": "Associate templates to OS, organization and location",
          "id": 45,
          "name": "associate",
          "selection": Array [
            Object {
              "label": "New",
              "value": "new",
            },
            Object {
              "label": "Never",
              "value": "never",
            },
            Object {
              "label": "Always",
              "value": "always",
            },
          ],
          "settingsType": "string",
          "value": "new",
        },
        Object {
          "description": "Should importing overwrite locked templates?",
          "id": 46,
          "name": "force",
          "settingsType": "bool",
          "value": false,
        },
      ]
    }
    resetField={[Function]}
    syncType="export"
    validationData={Object {}}
  />
</Form>
`;

exports[`NewTemplateSyncForm should render when for import settings 1`] = `
<Form
  disabled={false}
  errorTitle="Warning! "
  onCancel={[Function]}
  submitting={false}
>
  <SyncTypeRadios
    controlLabel="Action type"
    disabled={false}
    name="syncType"
    radios={
      Array [
        Object {
          "checked": true,
          "label": "Import",
          "onChange": [Function],
          "permission": "import",
          "value": "import",
        },
        Object {
          "checked": false,
          "label": "Export",
          "onChange": [Function],
          "permission": "export",
          "value": "export",
        },
      ]
    }
  />
  <SyncSettingsFields
    disabled={false}
    exportSettings={
      Array [
        Object {
          "description": "Import or export names matching this regex (case-insensitive; snippets are not filtered)",
          "id": 47,
          "name": "filter",
          "settingsType": "string",
          "value": "",
        },
        Object {
          "description": "Negate the filter for import/export",
          "id": 48,
          "name": "negate",
          "settingsType": "bool",
          "value": false,
        },
        Object {
          "description": "Target path to import and export. Different protocols can be used, for example /tmp/dir, git://example.com, https://example.com, ssh://example.com. When exporting to /tmp, note that production deployments may be configured to use private tmp.",
          "id": 49,
          "name": "repo",
          "settingsType": "string",
          "value": "https://github.com/theforeman/community-templates.git",
        },
      ]
    }
    importSettings={
      Array [
        Object {
          "description": "Associate templates to OS, organization and location",
          "id": 45,
          "name": "associate",
          "selection": Array [
            Object {
              "label": "New",
              "value": "new",
            },
            Object {
              "label": "Never",
              "value": "never",
            },
            Object {
              "label": "Always",
              "value": "always",
            },
          ],
          "settingsType": "string",
          "value": "new",
        },
        Object {
          "description": "Should importing overwrite locked templates?",
          "id": 46,
          "name": "force",
          "settingsType": "bool",
          "value": false,
        },
      ]
    }
    resetField={[Function]}
    syncType="import"
    validationData={Object {}}
  />
</Form>
`;
